File: lib/splitter/splitter_bar.dart
part of dock_spawn;

class SplitterBar {
  IDockContainer previousContainer; // The panel to the left/top side of the bar, depending on the bar orientation
  IDockContainer nextContainer;     // The panel to the right/bottom side of the bar, depending on the bar orientation
  DivElement barElement;
  bool stackedVertical;
  StreamSubscription<MouseEvent> mouseMovedHandler;
  StreamSubscription<MouseEvent> mouseDownHandler;
  StreamSubscription<MouseEvent> mouseUpHandler;
  MouseEvent previousMouseEvent;
  int minPanelSize = 50; // TODO: Get from container configuration
  Size previousContainerSize;
  Size nextContainerSize;
  
  SplitterBar(this.previousContainer, this.nextContainer, this.stackedVertical) {
    barElement = new DivElement();
    barElement.classes.add(stackedVertical ? "splitbar-horizontal" : "splitbar-vertical");
    
    mouseDownHandler = barElement.onMouseDown.listen(onMouseDown);
  }

  void onMouseDown(MouseEvent e) {
    _startDragging(e);
  }
  
  void onMouseUp(MouseEvent e) {
    _stopDragging(e);
  }
 
  bool readyToProcessNextDrag = true;
  void onMouseMoved(MouseEvent e) {
    if (!readyToProcessNextDrag) {
      print ("Skip");
      return;
    }
    readyToProcessNextDrag = false;
    var dockManager = previousContainer.dockManager;
    dockManager.suspendLayout();
    int dx = e.$dom_pageX - previousMouseEvent.$dom_pageX;
    int dy = e.$dom_pageY - previousMouseEvent.$dom_pageY;
    _performDrag(dx, dy);
    previousMouseEvent = e;
    readyToProcessNextDrag = true;
    dockManager.resumeLayout();
  }
  
  void _performDrag(int dx, int dy) {
    int previousWidth = previousContainer.containerElement.$dom_clientWidth;
    int previousHeight = previousContainer.containerElement.$dom_clientHeight;
    int nextWidth = nextContainer.containerElement.$dom_clientWidth;
    int nextHeight = nextContainer.containerElement.$dom_clientHeight;
    
    int previousPanelSize = stackedVertical ? previousHeight : previousWidth; 
    int nextPanelSize = stackedVertical ? nextHeight : nextWidth;
    int deltaMovement = stackedVertical ? dy : dx;
    int newPreviousPanelSize = previousPanelSize + deltaMovement; 
    int newNextPanelSize = nextPanelSize - deltaMovement;
    
    if (newPreviousPanelSize < minPanelSize || newNextPanelSize < minPanelSize) {
      // One of the panels is smaller than it should be.
      // In that case, check if the small panel's size is being increased
      bool continueProcessing = (newPreviousPanelSize < minPanelSize && newPreviousPanelSize > previousPanelSize) || 
          (newNextPanelSize < minPanelSize && newNextPanelSize > nextPanelSize);
      
      if (!continueProcessing) return;
    }

    if (stackedVertical) {
      previousContainer.resize(previousWidth, newPreviousPanelSize);
      nextContainer.resize(nextWidth, newNextPanelSize);
    } else {
      previousContainer.resize(newPreviousPanelSize, previousHeight);
      nextContainer.resize(newNextPanelSize, nextHeight);
    }
  }
  
  void _startDragging(MouseEvent e) {
    disableGlobalTextSelection();
    document.body.classes.add("disable-selection");
    if (mouseMovedHandler != null) {
      mouseMovedHandler.cancel();
      mouseMovedHandler = null;
    }
    if (mouseUpHandler != null) {
      mouseUpHandler.cancel();
      mouseUpHandler = null;
    }
    mouseMovedHandler = window.onMouseMove.listen(onMouseMoved);
    mouseUpHandler = window.onMouseUp.listen(onMouseUp);
    previousMouseEvent = e;
  }
  
  void _stopDragging(MouseEvent e) {
    enableGlobalTextSelection();
    document.body.classes.remove("disable-selection");
    if (mouseMovedHandler != null) {
      mouseMovedHandler.cancel();
      mouseMovedHandler = null;
    }
    if (mouseUpHandler != null) {
      mouseUpHandler.cancel();
      mouseUpHandler = null;
    }
  }
}

File: js/lib/splitter/SplitterBar.js
dockspawn.SplitterBar = function(previousContainer, nextContainer, stackedVertical)
{
    this.previousContainer = previousContainer; // The panel to the left/top side of the bar, depending on the bar orientation
    this.nextContainer = nextContainer;         // The panel to the right/bottom side of the bar, depending on the bar orientation
    this.stackedVertical = stackedVertical;
    this.barElement = document.createElement('div');
    this.barElement.classList.add(stackedVertical ? "splitbar-horizontal" : "splitbar-vertical");
    this.mouseDownHandler = new dockspawn.EventHandler(this.barElement, 'mousedown', this.onMouseDown.bind(this));
    this.minPanelSize = 50; // TODO: Get from container configuration
    this.readyToProcessNextDrag = true;
};

dockspawn.SplitterBar.prototype.onMouseDown = function(e)
{
    this._startDragging(e);
};

dockspawn.SplitterBar.prototype.onMouseUp = function(e)
{
    this._stopDragging(e);
};

dockspawn.SplitterBar.prototype.onMouseMoved = function(e)
{
    if (!this.readyToProcessNextDrag)
        return;
    this.readyToProcessNextDrag = false;

    var dockManager = this.previousContainer.dockManager;
    dockManager.suspendLayout();
    var dx = e.pageX - this.previousMouseEvent.pageX;
    var dy = e.pageY - this.previousMouseEvent.pageY;
    this._performDrag(dx, dy);
    this.previousMouseEvent = e;
    this.readyToProcessNextDrag = true;
    dockManager.resumeLayout();
};

dockspawn.SplitterBar.prototype._performDrag = function(dx, dy)
{
    var previousWidth = this.previousContainer.containerElement.clientWidth;
    var previousHeight = this.previousContainer.containerElement.clientHeight;
    var nextWidth = this.nextContainer.containerElement.clientWidth;
    var nextHeight = this.nextContainer.containerElement.clientHeight;

    var previousPanelSize = this.stackedVertical ? previousHeight : previousWidth;
    var nextPanelSize = this.stackedVertical ? nextHeight : nextWidth;
    var deltaMovement = this.stackedVertical ? dy : dx;
    var newPreviousPanelSize = previousPanelSize + deltaMovement;
    var newNextPanelSize = nextPanelSize - deltaMovement;

    if (newPreviousPanelSize < this.minPanelSize || newNextPanelSize < this.minPanelSize)
    {
        // One of the panels is smaller than it should be.
        // In that case, check if the small panel's size is being increased
        var continueProcessing = (newPreviousPanelSize < this.minPanelSize && newPreviousPanelSize > previousPanelSize) ||
            (newNextPanelSize < this.minPanelSize && newNextPanelSize > nextPanelSize);

        if (!continueProcessing)
            return;
    }

    if (this.stackedVertical)
    {
        this.previousContainer.resize(previousWidth, newPreviousPanelSize);
        this.nextContainer.resize(nextWidth, newNextPanelSize);
    }
    else
    {
        this.previousContainer.resize(newPreviousPanelSize, previousHeight);
        this.nextContainer.resize(newNextPanelSize, nextHeight);
    }
};

dockspawn.SplitterBar.prototype._startDragging = function(e)
{
    disableGlobalTextSelection();
    if (this.mouseMovedHandler)
    {
        this.mouseMovedHandler.cancel();
        delete this.mouseMovedHandler;
    }
    if (this.mouseUpHandler)
    {
        this.mouseUpHandler.cancel();
        delete this.mouseUpHandler;
    }
    this.mouseMovedHandler = new dockspawn.EventHandler(window, 'mousemove', this.onMouseMoved.bind(this));
    this.mouseUpHandler = new dockspawn.EventHandler(window, 'mouseup', this.onMouseUp.bind(this));
    this.previousMouseEvent = e;
};

dockspawn.SplitterBar.prototype._stopDragging = function(e)
{
    enableGlobalTextSelection();
    document.body.classList.remove("disable-selection");
    if (this.mouseMovedHandler)
    {
        this.mouseMovedHandler.cancel();
        delete this.mouseMovedHandler;
    }
    if (this.mouseUpHandler)
    {
        this.mouseUpHandler.cancel();
        delete this.mouseUpHandler;
    }
};
